Sources/MongoSwift/MongoClient.swift

```diff
@@ -252,41 +274,109 @@ public class MongoClient {
         self.decoder = BSONDecoder(options: options)
         self.sdamEventHandlers = []
         self.commandEventHandlers = []
+        self.state = .upAndRunning
+
         self.connectionPool.initializeMonitoring(client: self)
     }
 
     deinit {
-        assert(self.isClosed, "MongoClient was not closed before deinitialization")
+        assert(
+            self.state == .shutDown,
+            "MongoClient was not shut down before deinitialization. " +
+                "Please call `shutdown()` or `syncShutdown()` when the client is no longer needed."
+        )
     }
 
-    /// Shuts this `MongoClient` down, closing all connection to the server and cleaning up internal state.
-    /// Call this method exactly once when you are finished using the client. You must ensure that all operations
-    /// using the client have completed before calling this. The returned future must be fulfilled before the
-    /// `EventLoopGroup` provided to this client's constructor is shut down.
-    public func shutdown() -> EventLoopFuture<Void> {
+    /**
+     * Shuts this client down, closing all connections to the server and shutting down the underlying thread pool. If
+     * `.createNew` was specified as the `EventLoopGroupProvider`, this method will also shut down the underlying
+     * `EventLoopGroup`.
+     *
+     * You must call this method exactly once when you are finished using the client. You must ensure that all
+     * operations using the client have completed before calling this.
+     *
+     * If a `.shared` `EventLoopGroup` was provided to this client, it must not be shut down until after client
+     * shutdown has completed.
+     *
+     * Unlike other asynchronous methods in the driver, this method accepts a callback to execute upon completion of
+     * shutdown, rather than returning an `EventLoopFuture`. This is because, in cases where `.createNew` was
+     * specified, the underlying `EventLoopGroup` will be unusable once this method is called.
+     *
+     * - Parameters:
+     *   - queue: a `DispatchQueue` on which to execute the provided callback, defaulting to the global queue.
+     *   - callback: A callback to handle any errors that occur in the process of shutdown.
+     *
+     * If the error provided to the callback is non-nil, it is likely one of the following:
+     * - `LogicError` if `syncShutdown` or `shutdown` has already been called on this client.
+     */
+    public func shutdown(queue: DispatchQueue = .global(), _ callback: @escaping (Error?) -> Void) {
+        let error: Error? = self.stateLock.withLock {
```

Comment:
I think a deadlock would be possible if the elg was single threaded. If two concurrent tasks happened to call `shutdown` simultaneously and the context switched while one had the lock, blocking here in the thread without the lock would deadlock. Also, even if deadlock doesn't occur, we'd be blocking an entire event loop group thread rather than relying on its cooperative scheduling functionality.

Reply:
hmm... I'm not sure it's worth handling what happens when people call shutdown multiple times; we've explicitly stated you should call it once

Reply:
It's not restricted to just that case though. If any other thread holds the lock the same thing could occur. e.g. if an operation grabs it to check the state, `shutdown` is simultaneously called, and then the context switches to the `shutdown` thread. 

Regardless of all that though, I think the biggest problem is that blocking an event loop thread means it can't be used to run the rest of the things that are sharing the group.

Reply:
maybe I'm just confused about how event loops work, but how would the context switch in the middle of executing a single task?

Reply:
I think you're right, that doesn't seem to be possible. An event loop processes each task to completion before going on to the next one. I was thinking in an async/await mindset where the context switches on `await` (I think that's how this works), but that doesn't apply in Swift. The only scenario I can imagine that deadlocks is if we acquire the lock in one context and release it in the callback of another future, which we're not doing here.

Reply:
I don't think we have an issue here. If two concurrent tasks called `shutdown` simultaneously, one would have to wait for the lock. The lock is incredibly short-lived, and its not calling to any code which might reenter this method (which would indeed be a deadlock). I propose we move on here.

Reply:
You're right that deadlocks are not really an issue here, but again my main concern is blocking an event loop that we're sharing with the rest of the user's application. The lock is not always short-lived--remember shutdown isn't the only thing acquiring this lock. Every operation acquires it when it checks out a connection, and if the pool is depleted, the lock may be held for a while while the pending operation waits for connection to become available. While `shutdown` blocks on this lock, nothing on that event loop can progress. SwiftNIO's docs repeatedly mentions the importance of not blocking in event loop threads for this reason ([\[1\]](https://github.com/apple/swift-nio#a-note-on-blocking), [\[2\]](https://apple.github.io/swift-nio/docs/current/NIO/Protocols/EventLoop.html)). It would be okay to block in here if we knew the call was never made from an event loop thread, but that's what `syncShutdown` is for. In fact, the main distinction between `syncShutdown` and `shutdown` is that one is blocking and the other is non-blocking. If they both block, then I don't really see the purpose of having two methods.

 I strongly think we should not move on here.

Reply:
we call `tryCheckout()` to get a new connection, which returns nil immediately if one isn't available, and then the lock is released, so blocking until a connection is available is not an issue.

Reply:
If you look at the [implementation of `mongoc_client_pool_try_pop`](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/src/mongoc/mongoc-client-pool.c#L278), you can see that it still isn't immediate, since it blocks on the lock on the pool itself and also creates new clients if the pool is < maxPoolSize, which may block on the topology lock.